src/blog/wasm-tail-call.md

```diff
@@ -0,0 +1,165 @@
+
+When WebAssembly tail calls are enabled, Clang implements symmetric transfer as described in that blog post, but when tail calls are not enabled, Clang silently compiles the code without symmetric transfer, which could lead to stack overflows and is technically not a correct implementation of C++20!
+
+To see the difference in action, use Emscripten to compile the last example from the blog post linked above and observe that it only avoids overflowing the stack if tail calls are enabled. Note that due to a recently-fixed bug, this will only work correctly in Emscripten 3.1.35 or later.
+
+## Tail Calls in V8
+
+As we saw earlier, it is not the engine's responsibility to detect calls in tail position. This should be done upstream by the toolchain. So the only thing left to do for TurboFan (V8's optimizing compiler) is to emit an appropriate sequence of instructions based on the call kind and the target function signature.  For our fibonacci example from earlier, the stack would look like this:
+
+![Simple tail call in TurboFan](/_img/tail-call.png)
+
+On the left we are inside `fib_rec` (green), called by `fib` (blue) and about to recursively tail-call `fib_rec`. First we unwind the current frame by resetting the frame and stack pointer. The frame pointer just restores its previous value by reading it from the “Caller FP” slot. The stack pointer moves to the top of the parent frame, plus enough space for any potential stack parameters and stack return values for the callee (0 in this case, everything is passed by registers). Parameters are moved into their expected registers according to fib_rec's linkage (not shown in the diagram). And finally we start running `fib_rec`, which will start by creating a new frame.
+
+`fib_rec` will unwind and rewind itself like this until `n == 0`, at which point it will return `a` by register to `fib`.
+
+This is a simple case where all parameters and return values fit into registers, and the callee has the same signature as the caller. In the general case, we might need to do complex stack manipulations:
+
+- Read outgoing parameters from the old frame
+- Move parameters into the new frame
+- Adjust the frame size by moving the return address up or down, depending on the number of stack parameters in the callee
+
+All these reads and writes can conflict with each other, because we are reusing the same stack space. This is a crucial difference with a non-tail call, which would simply push all the stack parameters and the return address on top of the stack.
+
+![Complex tail call in TurboFan](/_img/tail-call-complex.png)
```

Comment:
Please export these images as SVG and then run them through https://jakearchibald.github.io/svgomg/ and use those instead of the PNGs.

Reply:
Done.